File: examples/junks/struct.go
package main

import (
	"fmt"
)

// Status : 
type Status string

const (
	// StatusOk : 
	StatusOk = Status("OK")
	// StatusNg : 
	StatusNg = Status("NG")
)

// String : stringer implementation
func (s Status) String() string {
	switch s {
	case StatusOk:
		return "ok"
	case StatusNg:
		return "ng"
	default:
		panic(fmt.Sprintf("unexpected Status %v, in string()", s))
	}

}
// ParseStatus : parse
func ParseStatus(s string) Status {
	switch s {
	case "OK":
		return StatusOk
	case "NG":
		return StatusNg
	default:
		panic(fmt.Sprintf("unexpected Status %v, in parse()", s))
	}

}

// Greeter : 
type Greeter interface {
	Greet() string
}


// MoreGreeter : hai
type MoreGreeter interface {
	Greeter
	Greet2() string
}


// Person : 
type Person struct {
	Name string // person's name
	Age int
	Father *Person
	Mother *Person
}


// MorePerson : 
type MorePerson struct {
	Person
	memo string
}

File: examples/junks/struct.py
from goaway import get_repository

r = get_repository()
f = r.package("main").file("main.go")

status = f.enum("Status", r.string)
with status as member:
    member("ok", "OK")
    member("ng", "NG")

with f.struct("Person") as field:
    field("Name", r.string, comment="person's name")
    field("Age", r.int)
    field("Father", f.structs["Person"].pointer)
    field("Mother", f.structs["Person"].pointer)

with f.struct("MorePerson") as field:
    field(f.structs["Person"])
    field("memo", r.string)

with f.interface("Greeter") as method:
    method("Greet", returns=r.string)

with f.interface("MoreGreeter", comment="hai") as method:
    method(f.interfaces["Greeter"])
    method("Greet2", returns=r.string)

# todo: embeded
print(r.writer.write(f, r.m))
